### PR TITLE
(CDAP-11411) (CDAP-11502) Container speedup and Spark2 in distributed mode

### PR DESCRIPTION
**Pull request by chtyim**
The major changes are as follow:

- Refactor DistributedProgramRunner so that there is on method of sub-class to setup the launch context
- Localize an expand program jar to each container so that the container doesn't redo the expansion of jar
- Use `spark.yarn.jar`/`spark.yarn.archive` configuration to have the Spark framework available on HDFS to save upload and localization time (if TWILL-232 is fixed, it could some extra 2 seconds, but that's a separate PR anyway)
- Detects and localize spark framework based on spark compat version.

### CODE REVIEW

File: cdap-app-fabric/src/main/java/co/cask/cdap/app/program/Programs.java [hunk 1]
```diff
@@ -60,6 +64,74 @@ public static Program create(CConfiguration cConf, @Nullable ProgramRunner progr
...
+    // The classloader should be ProgramClassLoader
+    Preconditions.checkArgument(classLoader instanceof ProgramClassLoader,
+                                "Program %s doesn't use ProgramClassLoader", originalProgram);
+
+    // The new program should be in the same namespace and app
+    ProgramId originalId = originalProgram.getId();
+    Preconditions.checkArgument(originalId.getNamespaceId().equals(programId.getNamespaceId()),
+                                "Program %s is not in the same namespace as %s", programId, originalId);
+    Preconditions.checkArgument(originalId.getParent().equals(programId.getParent()),
+                                "Program %s is not in the same namespace as %s", programId, originalId);
```
Comment (albertshau) [inline cdap-app-fabric/src/main/java/co/cask/cdap/app/program/Programs.java]: not in the same application

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/distributed/DistributedWorkerProgramRunner.java [hunk 2]
```diff
@@ -72,18 +74,21 @@ public ProgramController createProgramController(TwillController twillController
...
+  protected void setupLaunchConfig(LaunchConfig launchConfig, Program program, ProgramOptions options,
+                                   CConfiguration cConf, Configuration hConf, File tempDir) throws IOException {
+    ApplicationSpecification appSpec = program.getApplicationSpecification();
+    WorkerSpecification workerSpec = appSpec.getWorkers().get(program.getName());
 
-    String instances = programOptions.getArguments().getOption(ProgramOptionConstants.INSTANCES,
-                                                        String.valueOf(workerSpec.getInstances()));
-    Resources resources = SystemArguments.getResources(programOptions.getUserArguments(),
+    String instances = options.getArguments().getOption(ProgramOptionConstants.INSTANCES,
+                                                               String.valueOf(workerSpec.getInstances()));
```
Comment (albertshau) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/distributed/DistributedWorkerProgramRunner.java]: identation

File: cdap-spark-core-base/src/main/java/co/cask/cdap/app/runtime/spark/SparkPackageUtils.java [hunk 3]
```diff
@@ -56,111 +75,92 @@
...
+    switch (sparkCompat) {
+      case SPARK1_2_10:
+        LOCAL_SPARK_LIBRARIES.put(sparkCompat, Collections.singleton(getSpark1AssemblyJar(sparkLibrary, sparkHome)));
+        break;
+      case SPARK2_2_11:
+        LOCAL_SPARK_LIBRARIES.put(sparkCompat, getSpark2LibraryJars(sparkLibrary, sparkHome));
+        break;
+      default:
+        // This shouldn't happen
+        throw new IllegalStateException("Unsupport Spark version " + sparkCompat);
```
Comment (albertshau) [inline cdap-spark-core-base/src/main/java/co/cask/cdap/app/runtime/spark/SparkPackageUtils.java]: Unsupported

File: cdap-spark-core-base/src/main/java/co/cask/cdap/app/runtime/spark/SparkPackageUtils.java [hunk 4]
```diff
@@ -198,8 +198,81 @@ public static String prepareSparkResources(File tempDir, Map<String, LocalizeRes
...
+    } catch (IOException e) {
+      // Just log, don't throw.
+      // If we already located the Spark Assembly jar during visiting, we can still use the jar.
+      LOG.warn("Exception raised while inspecting {}", sparkLib, e);
+    }
+
+    Preconditions.checkState(!jar.isEmpty(), "Failed to locate Spark library from %s", sparkHome);
+
+    assemblyJar = jar.get(0);
+    LOG.info("Located Spark Assembly JAR in {}", assemblyJar);
```
Comment (albertshau) [inline cdap-spark-core-base/src/main/java/co/cask/cdap/app/runtime/spark/SparkPackageUtils.java]: should this be debug?

File: cdap-spark-core-base/src/main/java/co/cask/cdap/app/runtime/spark/SparkPackageUtils.java [hunk 5]
```diff
@@ -261,6 +357,188 @@ public static File locateSparkDefaultsConfFile(Map<String, String> env) {
...
+    SparkFramework framework = SPARK_FRAMEWORKS.get(sparkCompat);
+    if (framework != null && framework.getFrameworkLocation(lf).exists()) {
+      return framework;
+    }
+
+    Properties sparkConf = getSparkDefaultConf();
+    switch (sparkCompat) {
+      case SPARK1_2_10:
+        SPARK_FRAMEWORKS.put(sparkCompat, prepareSpark1Framework(sparkConf, lf));
+      break;
```
Comment (albertshau) [inline cdap-spark-core-base/src/main/java/co/cask/cdap/app/runtime/spark/SparkPackageUtils.java:390]: indentation

File: cdap-spark-core-base/src/main/java/co/cask/cdap/app/runtime/spark/SparkPackageUtils.java [hunk 6]
```diff
@@ -261,6 +357,188 @@ public static File locateSparkDefaultsConfFile(Map<String, String> env) {
...
+
+      case SPARK2_2_11:
+        SPARK_FRAMEWORKS.put(sparkCompat, prepareSpark2Framework(sparkConf, lf, tempDir));
+      break;
```
Comment (albertshau) [inline cdap-spark-core-base/src/main/java/co/cask/cdap/app/runtime/spark/SparkPackageUtils.java:394]: indentation

File: cdap-spark-core-base/src/main/java/co/cask/cdap/app/runtime/spark/distributed/SparkContainerLauncher.java [hunk 7]
```diff
@@ -44,28 +57,37 @@
...
     // Spark classes are in the system classloader which we want to rewrite.
-    URL[] classLoaderUrls = urls.toArray(new URL[urls.size()]);
-    ClassLoader classLoader = new SparkContainerClassLoader(classLoaderUrls, systemClassLoader.getParent());
+    ClassLoader classLoader = new SparkContainerClassLoader(urls.toArray(new URL[urls.size()]), parentClassLoader);
+
+    // Sets the context classloader and launch the actual Spark main class.
+    Thread.currentThread().setContextClassLoader(classLoader);
+
+    try {
+      Class<?> aClass = classLoader.loadClass("org.apache.spark.util.MutableURLClassLoader");
```
Comment (albertshau) [inline cdap-spark-core-base/src/main/java/co/cask/cdap/app/runtime/spark/distributed/SparkContainerLauncher.java]: is this leftover from debugging?

**Review by albertshau — COMMENTED** [review]
minor comments

File: cdap-app-fabric/src/main/java/co/cask/cdap/app/program/Programs.java (hunk 1 above)
Comment (chtyim) [reply cdap-app-fabric/src/main/java/co/cask/cdap/app/program/Programs.java]: fixed

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/distributed/DistributedWorkerProgramRunner.java (hunk 2 above)
Comment (chtyim) [reply cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/distributed/DistributedWorkerProgramRunner.java]: fixed

File: cdap-spark-core-base/src/main/java/co/cask/cdap/app/runtime/spark/SparkPackageUtils.java (hunk 3 above)
Comment (chtyim) [reply cdap-spark-core-base/src/main/java/co/cask/cdap/app/runtime/spark/SparkPackageUtils.java]: fixed

File: cdap-spark-core-base/src/main/java/co/cask/cdap/app/runtime/spark/SparkPackageUtils.java (hunk 4 above)
Comment (chtyim) [reply cdap-spark-core-base/src/main/java/co/cask/cdap/app/runtime/spark/SparkPackageUtils.java]: Sure. It was INFO.

File: cdap-spark-core-base/src/main/java/co/cask/cdap/app/runtime/spark/SparkPackageUtils.java (hunk 5 above)
Comment (chtyim) [reply cdap-spark-core-base/src/main/java/co/cask/cdap/app/runtime/spark/SparkPackageUtils.java:390]: Oh, I always have the `break` align with the `case` for each matching.

File: cdap-spark-core-base/src/main/java/co/cask/cdap/app/runtime/spark/distributed/SparkContainerLauncher.java (hunk 7 above)
Comment (chtyim) [reply cdap-spark-core-base/src/main/java/co/cask/cdap/app/runtime/spark/distributed/SparkContainerLauncher.java]: removed

**Review by albertshau — APPROVED** [review]
lgtm

**Review by albertshau — APPROVED** [review]
lgtm

**Comment by chtyim** [comment]
There is some permgen issue in unit-test.

**Comment by chtyim** [comment]
Identified the issue as Spark Streaming Thread Leakage that causes classloader leakage. It only affects unit-test and SDK, shouldn't affect distributed mode at all. Temporarily disabled spark streaming test and opened CDAP-11577 for fixing it.

**Comment by chtyim** [comment]
Build https://builds.cask.co/browse/CDAP-DUT5739

**Comment by chtyim** [comment]
Latest build https://builds.cask.co/browse/CDAP-DUT5739-22

**Review by albertshau — APPROVED** [review]
lgtm

**Comment by chtyim** [comment]
Build passed merging.

**Pull request merged**